File: src/golang/operation.go
package oresmiddleware

import (
	"context"
	"errors"
	"log/slog"
	"reflect"
)

type OperationTransport string

const (
	OperationTransportHTTP      OperationTransport = "http"
	OperationTransportTCP       OperationTransport = "tcp"
	OperationTransportWebSocket OperationTransport = "websocket"
)

type OperationScope string

const (
	OperationScopeRequest    OperationScope = "request"
	OperationScopeConnection OperationScope = "connection"
	OperationScopeMessage    OperationScope = "message"
	OperationScopeCallback   OperationScope = "callback"
)

type OperationFailureKind string

const (
	OperationFailureError            OperationFailureKind = "error"
	OperationFailurePanic            OperationFailureKind = "panic"
	OperationFailureCancelled        OperationFailureKind = "cancelled"
	OperationFailureDeadlineExceeded OperationFailureKind = "deadline_exceeded"
)

type OperationDescriptor struct {
	Transport OperationTransport
	Scope     OperationScope
	Name      string
}

// OperationFailure contains only bounded classification and correlation data.
// The panic value or returned error is never copied into this public value.
type OperationFailure struct {
	Kind      OperationFailureKind
	Code      string
	Transport OperationTransport
	Scope     OperationScope
	Operation string
	RequestID string
	TraceID   string
	ErrorType string
}

type OperationOutcome[T any] struct {
	Value   T
	Failure *OperationFailure
}

func (outcome OperationOutcome[T]) OK() bool { return outcome.Failure == nil }

type OperationFailureReporter func(context.Context, OperationFailure)

// CaptureOperationContext takes a defensive copy suitable for an explicitly
// propagated goroutine, TCP connection, WebSocket message, or queue callback.
func CaptureOperationContext(ctx context.Context) (RequestContext, bool) {
	return CurrentContext(ctx)
}

// RunOperationBoundary turns one callback into an isolated failure domain.
// Panics and returned errors become typed outcomes. Reporter panics are ignored
// so observability cannot replace the application/protocol outcome.
func RunOperationBoundary[T any](
	parent context.Context,
	requestContext RequestContext,
	descriptor OperationDescriptor,
	reporter OperationFailureReporter,
	operation func(context.Context) (T, error),
) (outcome OperationOutcome[T]) {
	if parent == nil {
		parent = context.Background()
	}
	ctx := WithRequestContext(parent, requestContext)

	defer func() {
		if recovered := recover(); recovered != nil {
			failure := newOperationFailure(
				OperationFailurePanic,
				descriptor,
				requestContext,
				safeErrorType(recovered, "panic"),
			)
			outcome = OperationOutcome[T]{Failure: &failure}
			reportOperationFailure(ctx, reporter, failure)
		}
	}()

	if failure, failed := contextFailure(ctx, descriptor, requestContext); failed {
		reportOperationFailure(ctx, reporter, failure)
		return OperationOutcome[T]{Failure: &failure}
	}

	value, err := operation(ctx)
	if err != nil {
		kind := OperationFailureError
		if errors.Is(err, context.DeadlineExceeded) {
			kind = OperationFailureDeadlineExceeded
		} else if errors.Is(err, context.Canceled) {
			kind = OperationFailureCancelled
		}
		failure := newOperationFailure(kind, descriptor, requestContext, safeErrorType(err, "error"))
		reportOperationFailure(ctx, reporter, failure)
		return OperationOutcome[T]{Failure: &failure}
	}

	if failure, failed := contextFailure(ctx, descriptor, requestContext); failed {
		reportOperationFailure(ctx, reporter, failure)
		return OperationOutcome[T]{Failure: &failure}
	}
	return OperationOutcome[T]{Value: value}
}

func contextFailure(
	ctx context.Context,
	descriptor OperationDescriptor,
	requestContext RequestContext,
) (OperationFailure, bool) {
	switch {
	case errors.Is(ctx.Err(), context.DeadlineExceeded):
		return newOperationFailure(OperationFailureDeadlineExceeded, descriptor, requestContext, "DeadlineExceeded"), true
	case errors.Is(ctx.Err(), context.Canceled):
		return newOperationFailure(OperationFailureCancelled, descriptor, requestContext, "Canceled"), true
	default:
		return OperationFailure{}, false
	}
}

func newOperationFailure(
	kind OperationFailureKind,
	descriptor OperationDescriptor,
	requestContext RequestContext,
	errorType string,
) OperationFailure {
	code := "operation_failed"
	switch kind {
	case OperationFailureDeadlineExceeded:
		code = "operation_deadline_exceeded"
	case OperationFailureCancelled:
		code = "operation_cancelled"
	case OperationFailurePanic:
		code = "operation_panicked"
	}
	return OperationFailure{
		Kind:      kind,
		Code:      code,
		Transport: descriptor.Transport,
		Scope:     descriptor.Scope,
		Operation: safeOperationName(descriptor.Name),
		RequestID: requestContext.RequestID,
		TraceID:   requestContext.TraceID,
		ErrorType: errorType,
	}
}

func safeOperationName(value string) string {
	if len(value) == 0 || len(value) > 128 {
		return "operation"
	}
	for _, item := range value {
		if !((item >= 'a' && item <= 'z') || (item >= 'A' && item <= 'Z') || (item >= '0' && item <= '9') || item == '_' || item == '-' || item == '.' || item == ':') {
			return "operation"
		}
	}
	return value
}

func safeErrorType(value any, fallback string) string {
	valueType := reflect.TypeOf(value)
	if valueType == nil {
		return fallback
	}
	for valueType.Kind() == reflect.Pointer {
		valueType = valueType.Elem()
	}
	name := valueType.Name()
	if name == "" || len(name) > 64 {
		return fallback
	}
	for _, item := range name {
		if !((item >= 'a' && item <= 'z') || (item >= 'A' && item <= 'Z') || (item >= '0' && item <= '9') || item == '_' || item == '-' || item == '.') {
			return fallback
		}
	}
	return name
}

func reportOperationFailure(
	ctx context.Context,
	reporter OperationFailureReporter,
	failure OperationFailure,
) {
	defer func() { _ = recover() }()
	if reporter != nil {
		reporter(ctx, failure)
		return
	}
	slog.ErrorContext(
		ctx,
		"operation failed",
		"operation.name", failure.Operation,
		"operation.transport", failure.Transport,
		"operation.scope", failure.Scope,
		"operation.outcome", failure.Kind,
		"error.type", failure.ErrorType,
		"request_id", failure.RequestID,
		"trace_id", failure.TraceID,
	)
}

// OresOperationFailureReporter emits a failure through the request logger
// installed by WrapWithOresLogger. Missing or failed transports remain fail-open.
func OresOperationFailureReporter(ctx context.Context, failure OperationFailure) {
	_ = Log(ctx).Error(
		"operation failed",
		map[string]any{
			"operation.name":      failure.Operation,
			"operation.transport": failure.Transport,
			"operation.scope":     failure.Scope,
			"operation.outcome":   failure.Kind,
			"error.type":          failure.ErrorType,
			"request.id":          failure.RequestID,
			"trace.id":            failure.TraceID,
		},
	)
}

File: src/operation.ts
import defaultLogger from "@oresoftware/next-loggers";
import {
  captureLogContext,
  runWithLogContext,
  type LogContext
} from "@oresoftware/next-loggers/context";

import {
  captureContext,
  runWithCapturedContext
} from "./context.js";
import type { RequestContext } from "./index.js";

export type OperationTransport = "http" | "tcp" | "websocket";
export type OperationScope = "request" | "connection" | "message" | "callback";
export type OperationFailureKind = "error" | "cancelled" | "deadline_exceeded";

export interface OperationDescriptor {
  transport: OperationTransport;
  scope: OperationScope;
  /** Stable low-cardinality name such as `orders.read` or `chat.message`. */
  name: string;
  /** Optional cooperative cancellation source owned by the protocol adapter. */
  signal?: AbortSignal;
}

export interface CapturedOperationContext {
  requestContext: RequestContext | undefined;
  /** Always present. An empty object represents explicitly captured absence. */
  logContext: LogContext;
}

/** Safe failure metadata. It deliberately excludes the exception message and payload. */
export interface OperationFailure {
  kind: OperationFailureKind;
  code: "operation_failed" | "operation_cancelled" | "operation_deadline_exceeded";
  transport: OperationTransport;
  scope: OperationScope;
  operation: string;
  requestId?: string;
  traceId?: string;
  errorType: string;
}

export interface OperationFailureEvent {
  failure: OperationFailure;
  /** Available only to an explicitly configured reporter; never copied to the public failure. */
  cause: unknown;
}

export type OperationFailureReporter = (
  event: OperationFailureEvent
) => void | Promise<void>;

export type OperationOutcome<T> =
  | { ok: true; value: T }
  | { ok: false; failure: OperationFailure };

export interface OperationBoundaryOptions {
  /** Defaults to a snapshot captured when `runOperationBoundary` is called. */
  context?: CapturedOperationContext;
  /** Defaults to the redacted, fail-open ores-otel reporter. */
  reportFailure?: OperationFailureReporter;
}

const safeToken = /^[A-Za-z0-9_.:-]{1,128}$/;

function safeOperationName(value: string): string {
  return safeToken.test(value) ? value : "operation";
}

function safeErrorType(cause: unknown): string {
  const candidate = cause instanceof Error
    ? cause.name
    : typeof cause;
  return safeToken.test(candidate) ? candidate : "UnknownError";
}

function failureKind(
  cause: unknown,
  signal: AbortSignal | undefined
): OperationFailureKind {
  const errorName = cause instanceof Error ? cause.name : "";
  const reasonName = signal?.reason instanceof Error ? signal.reason.name : "";
  if (errorName === "TimeoutError" || reasonName === "TimeoutError") {
    return "deadline_exceeded";
  }
  if (signal?.aborted || errorName === "AbortError") {
    return "cancelled";
  }
  return "error";
}

function failureCode(kind: OperationFailureKind): OperationFailure["code"] {
  switch (kind) {
    case "deadline_exceeded":
      return "operation_deadline_exceeded";
    case "cancelled":
      return "operation_cancelled";
    case "error":
      return "operation_failed";
  }
}

/** Captures both the middleware carrier and ores-otel's native log carrier. */
export function captureOperationContext(): CapturedOperationContext {
  return {
    requestContext: captureContext(),
    // An empty child frame is intentional: it prevents a callback captured
    // outside a request from inheriting whichever request invokes it later.
    logContext: captureLogContext() ?? {}
  };
}

/**
 * Builds a fresh immutable operation carrier from an explicit request context.
 * This is used by framework adapters before policy hooks run and again after
 * authentication enriches the actor/tenant fields. Only allow-listed values
 * become ambient log fields.
 */
export function operationContextFromRequestContext(
  context: RequestContext
): CapturedOperationContext {
  const fields: Record<string, unknown> = {
    "request.id": context.requestId,
    "trace.id": context.traceId,
    "request.started_at_unix_ms": context.startedAtUnixMs
  };
  if (context.spanId) fields["span.id"] = context.spanId;
  if (context.userId) fields["user.id"] = context.userId;
  if (context.tenantId) fields["tenant.id"] = context.tenantId;
  if (context.locale) fields["request.locale"] = context.locale;
  if (context.deadlineUnixMs !== undefined) {
    fields["request.deadline_unix_ms"] = context.deadlineUnixMs;
  }
  for (const [key, value] of Object.entries(context.baggage)) {
    if (key.startsWith("otel.")) fields[`baggage.${key}`] = value;
  }
  return { requestContext: context, logContext: { fields } };
}

/**
 * Re-enters both captured carriers. Explicitly absent frames mask unrelated
 * ambient state while the callback runs and are restored afterward.
 */
export function runWithCapturedOperationContext<T>(
  snapshot: CapturedOperationContext,
  operation: () => T
): T {
  return runWithLogContext(snapshot.logContext, () =>
    runWithCapturedContext(snapshot.requestContext, operation)
  );
}

/**
 * Default reporter: emits only bounded classification and correlation fields.
 * The raw cause is intentionally not serialized; applications may provide a
 * separately audited reporter when redacted stack capture is required.
 */
export const reportOresOperationFailure: OperationFailureReporter = async ({ failure }) => {
  try {
    await defaultLogger
      .error("operation failed")
      .addFields({
        "operation.name": failure.operation,
        "operation.transport": failure.transport,
        "operation.scope": failure.scope,
        "operation.outcome": failure.kind,
        "error.type": failure.errorType,
        ...(failure.requestId === undefined ? {} : { "request.id": failure.requestId }),
        ...(failure.traceId === undefined ? {} : { "trace.id": failure.traceId })
      })
      .send();
  } catch {
    // Logging/export must never replace an application or protocol outcome.
  }
};

async function reportSafely(
  reporter: OperationFailureReporter,
  event: OperationFailureEvent
): Promise<void> {
  try {
    await reporter(event);
  } catch {
    // User-supplied reporters remain fail-open for the guarded operation.
  }
}

/**
 * Executes one HTTP request, TCP connection/callback, or WebSocket message as
 * an isolated failure domain. Exceptions become typed outcomes; they do not
 * reject the event-loop callback or terminate the listener.
 *
 * Cancellation remains cooperative: the operation must observe the supplied
 * signal. The boundary classifies an already-aborted signal and AbortError /
 * TimeoutError failures but does not pretend JavaScript can force-stop a
 * promise that ignores cancellation.
 */
export async function runOperationBoundary<T>(
  descriptor: OperationDescriptor,
  operation: () => T | Promise<T>,
  options: OperationBoundaryOptions = {}
): Promise<OperationOutcome<T>> {
  const snapshot = options.context ?? captureOperationContext();
  const reporter = options.reportFailure ?? reportOresOperationFailure;
  const operationName = safeOperationName(descriptor.name);

  return runWithCapturedOperationContext(snapshot, async () => {
    try {
      if (descriptor.signal?.aborted) {
        throw descriptor.signal.reason ?? new DOMException("operation cancelled", "AbortError");
      }
      return { ok: true, value: await operation() };
    } catch (cause) {
      const kind = failureKind(cause, descriptor.signal);
      const failure: OperationFailure = {
        kind,
        code: failureCode(kind),
        transport: descriptor.transport,
        scope: descriptor.scope,
        operation: operationName,
        requestId: snapshot.requestContext?.requestId,
        traceId: snapshot.requestContext?.traceId,
        errorType: safeErrorType(cause)
      };
      await reportSafely(reporter, { failure, cause });
      return { ok: false, failure };
    }
  });
}

/** Captures context at registration time for event emitters and socket loops. */
export function bindOperationBoundary<Arguments extends unknown[], Result>(
  descriptor: OperationDescriptor,
  operation: (...arguments_: Arguments) => Result | Promise<Result>,
  options: Omit<OperationBoundaryOptions, "context"> = {}
): (...arguments_: Arguments) => Promise<OperationOutcome<Result>> {
  const context = captureOperationContext();
  return (...arguments_) =>
    runOperationBoundary(
      descriptor,
      () => operation(...arguments_),
      { ...options, context }
    );
}
